metrics/prometheus-metrics/src/main/scala/sttp/tapir/metrics/prometheus/PrometheusMetrics.scala

```diff
@@ -0,0 +1,136 @@
+  def withResponsesTotal: PrometheusMetrics[F] = copy(metrics = metrics + responsesTotal(registry))
+  def withResponsesDuration: PrometheusMetrics[F] = copy(metrics = metrics + responsesDuration(registry))
+  def withCustom(m: Metric[F, _]): PrometheusMetrics[F] = copy(metrics = metrics + m)
+
+  def metricsServerEndpoint: ServerEndpoint[Unit, Unit, CollectorRegistry, Any, F] =
+    metricsEp.serverLogic { _ => monad.unit(Right(registry)) }
+
+  def metricsInterceptor[B](ignoreEndpoints: Seq[Endpoint[_, _, _, _]] = Seq.empty): Interceptor[F, B] =
+    new MetricsInterceptor[F, B](metrics.toList, ignoreEndpoints :+ metricsEp)
+}
+
+object PrometheusMetrics {
+
+  implicit val schemaForCollectorRegistry: Schema[CollectorRegistry] = Schema.string[CollectorRegistry]
+
+  implicit val collectorRegistryCodec: Codec[String, CollectorRegistry, CodecFormat.TextPlain] =
+    Codec.anyStringCodec(TextPlain())(_ => DecodeResult.Value(new CollectorRegistry()))(r => {
+      val output = new StringWriter()
+      TextFormat.write004(output, r.metricFamilySamples)
+      output.close()
+      output.toString
+    })
+
+  def withDefaultMetrics[F[_]](implicit monad: MonadError[F]): PrometheusMetrics[F] =
+    PrometheusMetrics[F](
+      defaultRegistry,
+      Set(
+        requestsTotal(defaultRegistry),
+        requestsActive(defaultRegistry),
+        requestsFailures(defaultRegistry),
+        responsesTotal(defaultRegistry),
+        responsesDuration(defaultRegistry)
+      )
+    )
+
+  def requestsTotal[F[_]](registry: CollectorRegistry)(implicit monad: MonadError[F]): Metric[F, Counter] =
+    Metric[F, Counter](
+      Counter
+        .build()
+        .namespace("tapir")
+        .name("requests_total")
+        .help("Total HTTP requests")
+        .labelNames("path", "method")
+        .create()
+        .register(registry)
+    ).onRequest { (req, counter) => monad.unit(counter.labels(path(req), method(req)).inc()) }
+
+  def requestsActive[F[_]](registry: CollectorRegistry)(implicit monad: MonadError[F]): Metric[F, Gauge] =
+    Metric[F, Gauge](
+      Gauge
+        .build()
+        .namespace("tapir")
+        .name("requests_active")
+        .help("Active HTTP requests")
+        .labelNames("path", "method")
+        .create()
+        .register(registry)
+    ).onRequest { (req, gauge) => monad.unit(gauge.labels(path(req), method(req)).inc()) }
+      .onResponse { (req, _, gauge) => monad.unit(gauge.labels(path(req), method(req)).dec()) }
+
+  def requestsFailures[F[_]](registry: CollectorRegistry)(implicit monad: MonadError[F]): Metric[F, Counter] =
+    Metric[F, Counter](
+      Counter
+        .build()
+        .namespace("tapir")
+        .name("requests_failure")
+        .help("Unserved HTTP requests")
+        .labelNames("path", "method")
+        .register(registry)
+    ).onDecodeFailure { (req, counter) => monad.unit(counter.labels(path(req), method(req)).inc()) }
```

Comment:
decode failure is not a request failure. A single request can cause mulitple decode failures,until one endpoint matches and gives a decode success. So a failed request is rather one where the status code is non-2xx/3xx